File: BE/user-service/src/main/java/com/medcare/userservice/service/PrescriptionService.java
package com.medcare.userservice.service;

import com.medcare.userservice.dto.PrescriptionResponse;
import com.medcare.userservice.entity.PrescriptionStatus;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.util.List;

public interface PrescriptionService {
    PrescriptionResponse uploadPrescription(Long userId, MultipartFile file) throws IOException;
    List<PrescriptionResponse> getUserPrescriptions(Long userId);
    List<PrescriptionResponse> getAllPrescriptions();
    PrescriptionResponse getPrescriptionById(Long id, Long requesterId, String role);
    PrescriptionResponse updateStatus(Long id, PrescriptionStatus status, String note);
    PrescriptionResponse markAsUsed(Long id);
    List<PrescriptionResponse> getPendingPrescriptions();
    PrescriptionResponse analyzePrescription(Long id);
    PrescriptionResponse resetUsage(Long id);
    void deletePrescription(Long id);
    PrescriptionResponse updateExtractedData(Long id, String extractedData);
    PrescriptionResponse updatePrescriptionInfo(Long id, String hospitalName, String doctorName, String expiryDate);
    
    /**
     * Called by Order Service after successful payment.
     * Marks specific products as purchased in the prescription's extracted_data.
     */
    PrescriptionResponse fulfillMedicine(Long prescriptionId, List<Long> purchasedProductIds);
    PrescriptionResponse unfulfillMedicine(Long prescriptionId, List<Long> cancelledProductIds);
    
    /**
     * Called to validate if a product can be added to cart from this prescription.
     * Returns error message if not allowed, null if allowed.
     */
    String validateMedicineForCart(Long prescriptionId, Long productId);
}

File: FE/services/prescriptionService.ts
import axiosInstance from "@/lib/axiosInstance";
import { PrescriptionResponse, PrescriptionStatus } from "@/lib/types";

export const prescriptionService = {
  getMyPrescriptions: async (): Promise<PrescriptionResponse[]> => {
    return await axiosInstance.get("/user-service/api/users/prescriptions/me");
  },

  uploadPrescription: async (file: File): Promise<PrescriptionResponse> => {
    const formData = new FormData();
    formData.append("file", file);
    return await axiosInstance.post("/user-service/api/users/prescriptions/upload", formData, {
      headers: {
        "Content-Type": "multipart/form-data",
      },
    });
  },

  getPendingPrescriptions: async (): Promise<PrescriptionResponse[]> => {
    return await axiosInstance.get("/user-service/api/users/prescriptions/pending");
  },

  getAllPrescriptions: async (): Promise<PrescriptionResponse[]> => {
    // Note: This endpoint should be allowed for pharmacists
    return await axiosInstance.get("/user-service/api/users/prescriptions/all");
  },

  getPrescriptionById: async (id: number): Promise<PrescriptionResponse> => {
    return await axiosInstance.get(`/user-service/api/users/prescriptions/${id}`);
  },

  updatePrescriptionStatus: async (id: number, status: PrescriptionStatus, note?: string): Promise<PrescriptionResponse> => {
    return await axiosInstance.put(`/user-service/api/users/prescriptions/${id}/status`, null, {
      params: { status, note }
    });
  },

  analyzePrescription: async (id: number): Promise<PrescriptionResponse> => {
    return await axiosInstance.post(`/user-service/api/users/prescriptions/${id}/analyze`);
  },

  deletePrescription: async (id: number): Promise<void> => {
    return await axiosInstance.delete(`/user-service/api/users/prescriptions/${id}`);
  },
  
  updatePrescriptionInfo: async (id: number, info: { hospitalName?: string, doctorName?: string, expiryDate?: string }): Promise<PrescriptionResponse> => {
    return await axiosInstance.put(`/user-service/api/users/prescriptions/${id}/info`, null, {
      params: info
    });
  },

  updateExtractedData: async (id: number, extractedData: string): Promise<PrescriptionResponse> => {
    return await axiosInstance.post(`/user-service/api/users/prescriptions/${id}/update-extracted-data`, extractedData, {
      headers: { 'Content-Type': 'text/plain' }
    });
  },

  validateMedicineForCart: async (id: number, productId: number): Promise<void> => {
    await axiosInstance.get(`/user-service/api/users/prescriptions/${id}/validate`, {
      params: { productId }
    });
  }
};
